src/combinatorics/simple_graph/strongly_regular.lean

```diff
@@ -0,0 +1,147 @@
+
+## TODO
+- Define the complement of a graph
+- Prove that the complement of a strongly regular graph is strongly regular
+- Prove that the parameters of a strongly regular graph
+  obey relation `(n - k - 1) * m = k * (k - l - 1)`
+- Prove that if I is the identity matrix and J is the all-one matrix,
+  then the adj matrix A of SRG obeys relation A^2 = kI + lA + m(J - I - A)
+-/
+universes u
+
+namespace simple_graph
+variables {V : Type u}
+variables (G : simple_graph V)
+
+variables  [fintype V] [locally_finite G]
+
+/--
+A graph is strongly regular with parameters `n k l m` if
+ * its vertex set has cardinality `n`
+ * it is regular with degree `k`
+ * every pair of adjacent vertices has `l` common neighbors
+ * every pair of nonadjacent vertices has `m` common neighbors
+-/
+structure is_SRG_of (n k l m : ℕ) : Prop :=
+(card : fintype.card V = n)
+(regular : G.is_regular_of_degree k)
+(adj_common : ∀ (v w : V), G.adj v w → fintype.card (G.common_neighbors v w) = l)
+(nadj_common : ∀ (v w : V), ¬ G.adj v w → fintype.card (G.common_neighbors v w) = m)
+
+lemma degree_lt_card_verts (G : simple_graph V) (v : V) : G.degree v < fintype.card V :=
+begin
+  classical,
+  rw ← finset.card_univ,
+  rw degree,
+  apply finset.card_lt_card,
+  rw finset.ssubset_iff,
+  use v,
+  split,
+  { simp },
+  { apply finset.subset_univ },
+end
+
+lemma set.to_finset_subset_to_finset_iff (α : Type*) (s t : set α) [fintype s] [fintype t] :
+  s ⊆ t ↔ s.to_finset ⊆ t.to_finset :=
+⟨λ h x hx, set.mem_to_finset.mpr $ h $ set.mem_to_finset.mp hx,
+λ h x hx, set.mem_to_finset.mp $ h $ set.mem_to_finset.mpr hx⟩
+
+lemma card_common_neighbors_le_degree (v w : V) :
+  fintype.card (G.common_neighbors v w) ≤ G.degree v :=
+begin
+  rw degree,
+  have h := G.common_neighbors_subset_neighbor_set v w,
+  rw common_neighbors,
+  rw ← set.to_finset_card,
+  apply finset.card_le_of_subset,
+  rw common_neighbors at h,
+  rw neighbor_finset,
+  rw ← set.to_finset_subset_to_finset_iff _ _,
+  exact h,
+end
+
+lemma card_common_neighbors_lt_card_verts (G : simple_graph V) (v w : V) :
+  fintype.card (G.common_neighbors v w) < fintype.card V :=
+begin
+  classical,
+  have h := not_mem_left_common_neighbors G v w,
+  rw ← set.to_finset_card,
+  rw ← finset.card_univ,
+  apply finset.card_lt_card,
+  rw finset.ssubset_iff,
+  use v,
+  split,
+  { simp,
+    exact h },
+  { apply finset.subset_univ },
+end
+
+lemma adj_card_common_neighbors_lt_degree (v w : V) (h : G.adj v w) :
+  fintype.card (G.common_neighbors v w) < G.degree v :=
+begin
+  classical,
+  rw degree,
+  rw ← set.to_finset_card,
+  apply finset.card_lt_card,
+  rw finset.ssubset_iff,
+  use w,
+  split,
+  { rw set.mem_to_finset,
+    exact not_mem_right_common_neighbors G v w },
+  { rw finset.insert_subset,
+    split,
+    { simpa },
+    { rw neighbor_finset,
+      rw ← set.to_finset_subset_to_finset_iff,
+      exact common_neighbors_subset_neighbor_set G v w } },
+end
```

Comment:
I wasn't really able to compress this one, other than combining a few rewrites.  I'd suggest the following name for the lemma.
```lean
lemma card_common_neighbors_lt_degree_of_adj (v w : V) (h : G.adj v w) :
  fintype.card (G.common_neighbors v w) < G.degree v :=
begin
  classical,
  rw [degree, common_neighbors, ←set.to_finset_card],
  apply finset.card_lt_card,
  rw finset.ssubset_iff,
  use w,
  split,
  { rw set.mem_to_finset,
    apply not_mem_right_common_neighbors },
  { rw finset.insert_subset,
    split,
    { simpa, },
    { rw [neighbor_finset, ←set.subset_iff_to_finset_subset],
      apply common_neighbors_subset_neighbor_set } },
end
```